ci/scripts/lib/bazel_query.py

```diff
@@ -0,0 +1,115 @@
+        enables type confusion bugs to go unnoticed. For example, a tuple's
+        string representation, complete with parentheses and commas, can
+        accidentally be inserted into a target name via `string.format()`.
+
+        [0]: https://bazel.build/concepts/labels#target-names
+
+        Yields:
+          A tuple for each target that has banned characters in its name. The
+          first element is the Bazel target's name. The second element is the
+          set of characters that must be removed from the target's name.
+
+        """
+        allowed_chars = set(string.ascii_letters + string.digits + '/:_-.')
+        for target in self.query("//..."):
+            if bad_chars := set(target) - allowed_chars:
+                yield (target, bad_chars)
+
+    def find_empty_test_suites(self) -> Iterable[str]:
+        """Find test_suite targets in //... that contain zero tests.
+
+        Finding empty test suites is not as simple as querying for test_suite
+        targets where `tests = []`. In fact, that is a special case of
+        test_suite that causes it to select all tests in the package.
+
+        There are a few ways to wind up with an empty test suite. One way is to
+        provide a nonempty `tests` argument containing only nonexistent labels.
+        Another way is to provide a non-empty list of tests that exist, but to
+        specify `tags` that exclude all the tests.
+
+        Yields:
+          Names of Bazel targets.
+        """
+        def print_progress(i, n):
+            if n // 5 == 0 or i % (n // 5) == 0:
+                print(self.find_empty_test_suites.__name__, end="")
+                print(": Running query {}/{}...".format(i + 1, n))
+
+        query_test_suites = BazelQuery.rule_exact("test_suite", "//...")
+        all_test_suites = self.query(query_test_suites)
+
+        for i, test_suite in enumerate(all_test_suites):
+            print_progress(i, len(all_test_suites))
+            tests = self.query("tests({})".format(test_suite))
+            if len(tests) == 0:
+                yield test_suite
+
+    def find_non_manual_test_suites(self) -> Iterable[str]:
+        """Find test_suite targets in //... that are not tagged with 'manual'."""
+        query_pieces = [
+            BazelQuery.rule_exact("test_suite", "//..."),
+            "except",
+            BazelQuery.tag_exact("manual", "//..."),
+        ]
+        query_str = " ".join(query_pieces)
+        return self.query(query_str)
+
+    def query(self, query: str) -> List[str]:
+        """Perform a Bazel query and return the resulting targets."""
+        if self._backend:
+            return self._backend(query)
+
+        bazel = subprocess.run(["./bazelisk.sh", "query", query],
+                               stdout=subprocess.PIPE,
+                               stderr=subprocess.DEVNULL,
+                               encoding='utf-8',
+                               check=True)
```

Comment:
You should probably specify --output here to ensure it's not set in a .bazelrc file